File: python/examples/error_handling.py
#!/usr/bin/env python3
"""
Error Handling Best Practices

This example demonstrates:
1. Proper exception handling
2. Retry strategies
3. Timeout handling
4. Validation errors
5. Network error recovery
"""

import asyncio
import logging
import os
from kaleidoswap_sdk.client import KaleidoClient
from kaleidoswap_sdk.exceptions import (
    NetworkError,
    AuthenticationError,
    RateLimitError,
    ValidationError,
)

# Configure logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)

# Configuration
API_URL = os.getenv("KALEIDO_API_URL", "https://api.staging.kaleidoswap.com")
API_KEY = os.getenv("KALEIDO_API_KEY", "invalid_key_for_demo")


async def example_authentication_error():
    """Demonstrate handling authentication errors."""
    logger.info("\n🔐 Example 1: Authentication Error Handling")

    # Use an invalid API key
    client = KaleidoClient(base_url=API_URL, api_key="invalid_key_123")

    try:
        await client.list_assets()
    except AuthenticationError as e:
        logger.warning(f"  ⚠️  Authentication failed (expected): {e}")
        logger.info("  💡 Solution: Check your API key and ensure it's valid")
    except Exception as e:
        logger.info(
            f"  Note: Got {type(e).__name__} instead (API might not require auth)"
        )
    finally:
        await client.close()


async def example_validation_error():
    """Demonstrate handling validation errors."""
    logger.info("\n✅ Example 2: Validation Error Handling")

    client = KaleidoClient(base_url=API_URL)

    try:
        # Try to get a quote with invalid parameters
        await client.get_quote(
            from_asset="INVALID_ASSET",
            to_asset="ALSO_INVALID",
            from_amount=-1000,  # Invalid negative amount
        )
    except ValidationError as e:
        logger.warning(f"  ⚠️  Validation error (expected): {e}")
        logger.info("  💡 Solution: Validate input parameters before making API calls")
    except Exception as e:
        logger.info(f"  Got error: {type(e).__name__}: {e}")
    finally:
        await client.close()


async def example_timeout_handling():
    """Demonstrate handling timeout errors."""
    logger.info("\n⏱️  Example 3: Timeout Error Handling")

    client = KaleidoClient(
        base_url=API_URL,
        timeout=0.001,  # Very short timeout to trigger error
    )

    try:
        await client.list_assets()
    except NetworkError as e:
        logger.warning(f"  ⚠️  Network/timeout error (expected): {e}")
        logger.info("  💡 Solution: Increase timeout or check network connectivity")
    except Exception as e:
        logger.info(f"  Got error: {type(e).__name__}: {e}")
    finally:
        await client.close()


async def example_retry_strategy():
    """Demonstrate retry strategy for transient errors."""
    logger.info("\n🔄 Example 4: Retry Strategy")

    client = KaleidoClient(base_url=API_URL)

    max_retries = 3
    retry_delay = 1.0

    for attempt in range(max_retries):
        try:
            logger.info(f"  Attempt {attempt + 1}/{max_retries}...")
            assets = await client.list_assets()
            logger.info(f"  ✅ Success! Found {len(assets.assets)} assets")
            break
        except NetworkError as e:
            if attempt < max_retries - 1:
                logger.warning(f"  ⚠️  Network error, retrying in {retry_delay}s: {e}")
                await asyncio.sleep(retry_delay)
                retry_delay *= 2  # Exponential backoff
            else:
                logger.error(f"  ❌ Failed after {max_retries} attempts: {e}")
        except Exception as e:
            logger.error(f"  ❌ Unexpected error: {e}")
            break

    await client.close()


async def example_rate_limit_handling():
    """Demonstrate handling rate limit errors."""
    logger.info("\n🚦 Example 5: Rate Limit Handling")

    client = KaleidoClient(base_url=API_URL)

    try:
        # Make many requests quickly
        for i in range(5):
            logger.info(f"  Request {i + 1}...")
            await client.list_assets()
            await asyncio.sleep(0.1)

        logger.info("  ✅ All requests succeeded")

    except RateLimitError as e:
        logger.warning(f"  ⚠️  Rate limit exceeded: {e}")
        logger.info(
            "  💡 Solution: Implement request throttling or wait before retrying"
        )
    except Exception as e:
        logger.info(f"  Note: {type(e).__name__}: {e}")
    finally:
        await client.close()


async def example_swap_timeout_handling():
    """Demonstrate handling swap timeout errors."""
    logger.info("\n⏳ Example 6: Swap Timeout Handling")

    client = KaleidoClient(
        base_url="http://localhost:8000", node_url="http://localhost:3001"
    )

    try:
        # Try to wait for a non-existent swap with short timeout
        await client.wait_for_swap_completion(
            payment_hash="nonexistent_hash_12345",
            timeout=5.0,  # 5 second timeout
            poll_interval=1.0,
        )
    except TimeoutError as e:
        logger.warning(f"  ⚠️  Swap timeout (expected): {e}")
        logger.info("  💡 Solution: Increase timeout or check swap status manually")
    except Exception as e:
        logger.info(f"  Got error: {type(e).__name__}: {e}")
    finally:
        await client.close()


async def example_proper_cleanup():
    """Demonstrate proper resource cleanup."""
    logger.info("\n🧹 Example 7: Proper Resource Cleanup")

    client = None
    try:
        client = KaleidoClient(base_url=API_URL)
        assets = await client.list_assets()
        logger.info(f"  ✅ Got {len(assets.assets)} assets")
    except Exception as e:
        logger.error(f"  ❌ Error: {e}")
    finally:
        if client:
            await client.close()
            logger.info("  ✅ Client properly closed")


async def example_context_manager():
    """Demonstrate using context manager for automatic cleanup."""
    logger.info("\n🎯 Example 8: Using Context Manager (Best Practice)")

    # Using async context manager ensures proper cleanup
    async with KaleidoClient(base_url=API_URL) as client:
        try:
            assets = await client.list_assets()
            logger.info(f"  ✅ Got {len(assets.assets)} assets")
        except Exception as e:
            logger.error(f"  ❌ Error: {e}")

    logger.info("  ✅ Client automatically closed by context manager")


async def main():
    """Run all error handling examples."""
    logger.info("=" * 60)
    logger.info("Error Handling Best Practices Examples")
    logger.info("=" * 60)

    try:
        await example_authentication_error()
        await example_validation_error()
        await example_timeout_handling()
        await example_retry_strategy()
        await example_rate_limit_handling()
        await example_swap_timeout_handling()
        await example_proper_cleanup()
        await example_context_manager()

        logger.info("\n" + "=" * 60)
        logger.info("✅ All error handling examples completed!")
        logger.info("=" * 60)

    except Exception as e:
        logger.error(f"❌ Unexpected error in main: {e}", exc_info=True)


if __name__ == "__main__":
    asyncio.run(main())

File: bindings/deprecated-typescript/examples/error-handling.ts
/**
 * Error Handling Best Practices
 * 
 * This example demonstrates:
 * 1. Proper exception handling
 * 2. Input validation
 * 3. Retry strategies
 * 4. Graceful degradation
 */

import { KaleidoClient, KaleidoError } from '../src';

const API_URL = process.env.KALEIDO_API_URL || 'https://api.staging.kaleidoswap.com';

async function exampleBasicErrorHandling() {
    console.log('\n🔐 Example 1: Basic Error Handling');

    const client = new KaleidoClient({ baseUrl: API_URL });

    try {
        const assets = await client.listAssets();
        const assetsList = typeof assets === 'string' ? JSON.parse(assets) : assets;
        console.log(`  ✅ Found ${assetsList.length} assets`);
    } catch (e) {
        if (e instanceof KaleidoError) {
            console.log(`  ⚠️  Kaleido API error: ${e.message}`);
        } else {
            console.log(`  ❌ Unexpected error: ${e}`);
        }
    }
}

async function exampleInputValidation() {
    console.log('\n✅ Example 2: Input Validation');

    const client = new KaleidoClient({ baseUrl: API_URL });

    // Validate negative amount
    console.log('  Testing negative amount...');
    const validation1 = await client.validateAmount(-100, 'BTC');
    if (!validation1.valid) {
        console.log(`    ❌ Invalid: ${validation1.errors.join(', ')}`);
    } else {
        console.log('    ✅ Valid (unexpected)');
    }

    // Validate zero amount
    console.log('  Testing zero amount...');
    const validation2 = await client.validateAmount(0, 'BTC');
    if (!validation2.valid) {
        console.log(`    ❌ Invalid: ${validation2.errors.join(', ')}`);
    } else {
        console.log('    ✅ Valid (unexpected)');
    }

    // Validate valid amount
    console.log('  Testing valid amount (0.01 BTC)...');
    try {
        const validation3 = await client.validateAmount(0.01, 'BTC');
        if (validation3.valid) {
            console.log(`    ✅ Valid: ${validation3.rawAmount} atomic units`);
        } else {
            console.log(`    ❌ Invalid: ${validation3.errors.join(', ')}`);
        }
    } catch (e) {
        console.log(`    Could not validate: ${e}`);
    }
}

async function exampleRetryStrategy() {
    console.log('\n🔄 Example 3: Retry Strategy');

    const client = new KaleidoClient({ baseUrl: API_URL });
    const maxRetries = 3;
    let retryDelay = 1000;

    for (let attempt = 1; attempt <= maxRetries; attempt++) {
        try {
            console.log(`  Attempt ${attempt}/${maxRetries}...`);
            const assets = await client.listAssets();
            const assetsList = typeof assets === 'string' ? JSON.parse(assets) : assets;
            console.log(`  ✅ Success! Found ${assetsList.length} assets`);
            break;
        } catch (e) {
            if (attempt < maxRetries) {
                console.log(`  ⚠️  Error, retrying in ${retryDelay}ms: ${e}`);
                await new Promise(resolve => setTimeout(resolve, retryDelay));
                retryDelay *= 2; // Exponential backoff
            } else {
                console.log(`  ❌ Failed after ${maxRetries} attempts: ${e}`);
            }
        }
    }
}

async function exampleInvalidPairTicker() {
    console.log('\n📊 Example 4: Invalid Pair Ticker Handling');

    const client = new KaleidoClient({
        baseUrl: API_URL,
        nodeUrl: 'http://localhost:3001',
    });

    // Test canTrade with non-existent pair
    const canTrade = await client.canTrade('FAKE', 'COIN');
    console.log(`  Can trade FAKE/COIN: ${canTrade} (expected: false)`);
}

async function exampleGracefulDegradation() {
    console.log('\n🛡️  Example 5: Graceful Degradation');

    // Client without node
    const clientNoNode = new KaleidoClient({ baseUrl: API_URL });

    if (clientNoNode.hasNode()) {
        console.log('  ✅ Node is configured');
    } else {
        console.log('  ℹ️  No node configured - operating in API-only mode');
        console.log('    Some features like whitelistTrade won\'t be available');
    }

    // Client with node
    const clientWithNode = new KaleidoClient({
        baseUrl: API_URL,
        nodeUrl: 'http://localhost:3001',
    });

    if (clientWithNode.hasNode()) {
        console.log('  ✅ Node is configured (with nodeUrl)');
    }
}

async function main() {
    console.log('='.repeat(60));
    console.log('Error Handling Best Practices Examples');
    console.log('='.repeat(60));

    try {
        await exampleBasicErrorHandling();
        await exampleInputValidation();
        await exampleRetryStrategy();
        await exampleInvalidPairTicker();
        await exampleGracefulDegradation();

        console.log('\n' + '='.repeat(60));
        console.log('✅ All error handling examples completed!');
        console.log('='.repeat(60));
    } catch (e) {
        console.error('❌ Unexpected error in main:', e);
    }
}

main();
